[RFC] Windows :terminal support

erw7 opened the pull request:
Implementation based on the work of equalsraf. Changes from the equalsraf work  are as follows.

* Changed uv\_run to be called in pty\_process\_spawn because uv\_connect\_cb was not called before calling uv\_read\_start that caused assert.
* Changed to poll using timer until uv\_is\_readable returns false because it does not stop process output polling until winpty-agent is finished.
* Implemented emulate quotiong of libuv's arguments to build cmdline.
* Changed by changing return value of pty\_process\_spawn.
* Changed to set proc->pid.
* Changed enable some test on windows.
* Changed to use utf16_to_utf8 function implemented in mbyte.c.

We do not think that the place to quote the command line is optimal, so we need to consider it.

If the last sleep of tty-test.c is enabled, the following error occurs in the functionaltest.

```
-- Output to stderr:
Uncaught Error: attempt to call a number value
stack traceback:
        [C]: at 0x67191e2e
```

In the case of Windows, I disabled it, is it really necessary to sleep for 100 seconds even in other environments?

At the moment there are the following problems.

<ul>
<li><del>Nvim will crash in the next step.</del>
<ul>
<li><del>nvim-qt.exe -- -u NONE</del></li>
<li><del>:terminal</del></li>
<li><del>:q</del></li>
</ul>
<del>There is no problem in the case of the following step.</del>
<ul>
<li><del>nvim.exe -u NONE</del></li>
<li><del>nvim-qt.exe --server \\.\pipe\nvim-xxxx-0</del></li>
<li><del>:terminal</del></li>
<li><del>:q</del></li>
</ul>
<del>In this case the errorlevel is also 0.</del></li>
<li><del>In the following procedure E301: Oops, lost the swap file !!! error will occur.</del>
<ul>
<li><del>nvim-qt.exe -- -u NONE</del></li>
<li><del>:e test</del></li>
<li><del>:call termopen("dir")</del></li>
</ul>
</ul>

rprichard committed:
FIXME: Checkpoint a neovim<->winpty integration.

Handling of process exit is still broken.  It detects the moment when the
child process exits, then quickly stops polling for process output.  It
should continue polling for output until the agent has scraped all of the
process' output.  This problem is easy to noticed by running a command like
"dir && exit", but even typing "exit<ENTER>" can manifest the problem --
the "t" might not appear.

winpty's Cygwin adapter handles shutdown by waiting for the agent to close
the CONOUT pipe, which it does after it has scraped the child's last
output.  AFAIK, neovim doesn't do anything interesting when winpty closes
the CONOUT pipe.

Commit:
Install winpty-agent.exe with nvim

erw7 committed:
Windows :terminal support

equalsraf commented:
@erw7  you can remove 2c844792c48511ae5f84ef32bc6979384b0b39f1 and 74b6348528c4fed564ab011701e517ce14bc8e3e they are already in the master branch.

justinmk commented:
Can you rebase this on master?

equalsraf commented:
> Can you rebase this on master?

Some conflicts but the rebase works away that merge commit, see f877eb67e36271514f62e00999e9a922206d08b7

erw7 commented:
How can I remove it.

brcolow commented:
To remove the commits do:

`git rebase -i HEAD~7` which will open up the editor you have configured with git.

Now change the word `pick` before the 2c84479 and 74b6348 commits to `drop`, write the file and quit. Then do `git push --force`. At least, I think that's what you should do - not a git aficionado by any means and I frequently mess up.

erw7 commented:
Thank you brcolow.
I tried it. 08661d3 to 439c203 are entering, but is this OK? Should I further rebase and 08661d3 to 439c203 also need to drop?

justinmk commented:
`git rebase -i upstream/master`

Or origin/master

erw7 commented:
I tried it, but this time 8f0aa5d has disappeared. I thought about going back with git reflog, but once I deleted all the local files, it does not remain. May I do as follows

```
git checkout -b patch-windows-terminal2  equalsraf/patch-windows-terminal
git pull -f origin patch-windows-terminal
```

justinmk commented:
With your current branch, just do `git cherry-pick 8f0aa5d` and it will bring that commit over.

erw7 commented:
I tried it, but it becomes as follows.

```
>git cherry-pick 8f0aa5df
The previous cherry-pick is now empty, possibly due to conflict resolution.
If you wish to commit it anyway, use:

    git commit --allow-empty

Otherwise, please use 'git reset'
On branch patch-windows-terminal
Your branch is up-to-date with 'origin/patch-windows-terminal'.
You are currently cherry-picking commit 8f0aa5df.

nothing to commit, working tree clean
>git status
On branch patch-windows-terminal
Your branch is up-to-date with 'origin/patch-windows-terminal'.
You are currently cherry-picking commit 8f0aa5df.
  (all conflicts fixed: run "git cherry-pick --continue")
  (use "git cherry-pick --abort" to cancel the cherry-pick operation)

nothing to commit, working tree clean
```

I created another branch and tried ```git commit --allow-empty```, 8f0aa5df came back, but it was empty.

I am sorry for taking care of it.

erw7 commented:
After canceling the same change at 57b1dcd ```git cherry-pick 8f0aa5d```, 8f0aa5d returned.

erw7 committed:
Fixed problems with jobstart and jobstop

Changing to make sure that proc->in, proc-> is not NULL, because nvim
crashed when starting a job with pty.

Changing to make sure that proc->out is not NULL, because nvim crashed
when stopping a job opened with pty.

erw7 committed:
Fix to fail build on *nix

Remove pty_process_win.h from NVIM_HEADERS on *nix.

erw7 committed:
Add boundary check to quote_cmd_arg

erw7 committed:
Fix clint issues

erw7 committed:
Fix Travis CI issue

erw7 committed:
Changed to output screen size with <C-q> key

tty-test.exe causes abnormal termination with low repeatability, try
changing it so as not to use SIGWINCH.

justinmk commented:
This works. Running `vim` in `bash --login -i` from git-for-windows 1.8.3 has messed up colors, but this is a good start!

erw7 commented:
What if you start vim as follows?

```
bash --login -i
export VIM=/usr/share/vim
export VIMRUNTIME=/usr/share/vim/vim80
vim
```

Anyway, i finished this work.

justinmk commented:
The vim colors issue was bogus, same issue happens in normal cmd.exe (I didn't think to check there because I normally run a different version of vim there, not the old git-for-windows vim). 

Thanks for working on this!

erw7 commented:
There are many points that do not follow the coding guide. I'll fix them.

Please remove RFC label temporarily.

ZyX-I commented:
@erw7 Labels are managed by bot, based on the PR title.

erw7 committed:
Change to follow coding guide

erw7 committed:
Add error log output

erw7 committed:
Change return value of pty_process_win

justinmk commented on src/nvim/os/pty_process_win.c:
```diff
@@ -102,6 +107,7 @@ int pty_process_spawn(PtyProcess *ptyproc)
       NULL,  // Optional environment variables
       &err);
   if (spawncfg == NULL) {
+    write_winpty_elog("Faied winpty_spawn_config_new.", &status, &err);
```
typo

justinmk commented on src/nvim/os/pty_process_win.c:
```diff
@@ -182,6 +195,19 @@ static void pty_process_connect_cb(uv_connect_t *req, int status)
+  FUNC_ATTR_NONNULL_ALL
+{
+  PtyProcess *ptyproc =
+    (PtyProcess *)((uv_handle_t *)wait_eof_timer->data);
```
we don't cast `void *` pointers, instead just:

    PtyProcess *ptyproc = wait_eof_timer->data;

justinmk commented on src/nvim/os/pty_process_win.c:
```diff
@@ -88,54 +82,73 @@ int pty_process_spawn(PtyProcess *ptyproc)
+                    &process_handle,
+                    NULL,  // Optional thread handle
+                    NULL,  // Optional create process error
+                    &err)) {
```
If I'm not mistaken, line 112 should be `// Optional environment variables`. The next parameter is for errors.

erw7 committed:
Fix the issue that reviewer pointed out

justinmk commented on test/functional/fixtures/tty-test.c:
```diff
@@ -9,10 +9,12 @@ uv_tty_t tty;
-  DWORD dwProcessId;
-  GetWindowThreadProcessId(consoleWnd, &dwProcessId);
-  return GetCurrentProcessId() == dwProcessId;
+  // XXX: We need to make proper detect owns tty
```
is this planned for this PR, or later?

erw7 replied:
I do not plan on this PR.
I do not understand what to check here on Windows.

justinmk commented on test/functional/fixtures/tty-test.c:
```diff
@@ -72,21 +76,32 @@ static void read_cb(uv_stream_t *stream, ssize_t cnt, const uv_buf_t *buf)
     if (buf->base[i] == 3) {
       (*interrupted)++;
+    } else if (buf->base[i] == 17) {
+      prsz = true;
```
could you add a comment mentioning where 17 comes from? Or make it a `#define`.

justinmk commented on test/functional/terminal/helpers.lua:
```diff
@@ -30,10 +30,13 @@ local function clear_attrs() feed_termcode('[0;10m') end
 local function enable_mouse() feed_termcode('[?1002h') end
 local function disable_mouse() feed_termcode('[?1002l') end
+local function print_screen_size()
+  helpers.sleep(1000)
```
1s sleep in the tests is pretty brutal, and this is used in many tests... can you use `helpers.retry()` instead? See `test/functional/autocmd/termclose_spec.lua` for example.

Note: retry() depends on an error being thrown--it won't retry on test "failure". (Again see `test/functional/autocmd/termclose_spec.lua`)

erw7 replied:
I defined CTRL_Q, and changed to use retray instead of wait.

justinmk commented on third-party/CMakeLists.txt:
```diff
@@ -87,8 +87,8 @@ endif()
-set(LIBUV_URL https://github.com/libuv/libuv/archive/v1.9.1.tar.gz)
-set(LIBUV_SHA256 a6ca9f0648973d1463f46b495ce546ddcbe7cce2f04b32e802a15539e46c57ad)
+set(LIBUV_URL https://github.com/libuv/libuv/archive/v1.11.0.tar.gz)
+set(LIBUV_SHA256 6ec7eec6ecc24b1a8ffedebedb2fe9313fffb5410de89aaf784dd01080411c7a)
```
Not opposed to this, but curious: is it required, and if so, why?

erw7 replied:
Although the intention of equalsraf of the original commit is unknown, as far as I tried, ```jobstop( jobstart('cmd',{'pty':1}))```  in version 1.9.1 caused assert on line 159 of libuv/src/win/handle-intl.h.
I have not confirmed which version was fixed.

erw7 replied:
The problem of nvim crashing when exiting nvim without exiting the terminal was also solved by updating the libuv version.

erw7 replied:
The problem of nvim crashing was irrelevant to the version of libuv. nvim -u NORC does not crash.

erw7 committed:
Add definition of CTRL_Q, CTRL_C

erw7 committed:
Change to retry from wait

erw7 committed:
Improve error code translation

erw7 committed:
Remove write_winpty_elog and write_elog function

erw7 committed:
Change to not quoting for cmd.exe

justinmk commented:
Would like to merge this as soon as we release 0.2. Can't risk the libuv upgrade before then.

erw7 commented:
Although there is no problem locally, I am worried that the test will fail with CI. But I agree to your proposal.

jalvesaq commented:
I tested this pull request on Linux (Ubuntu 17.04) today and did not note any issue.

justinmk commented:
This may also fix #6672 and #6143.

erw7 commented:
#6143 is caused due to casting PtyProcess to LibuvProcess in pty_process_win.h. Therefore, this will fix #6143. #6672 is not reproduced in my environment, so it is unknown whether this will be fixed.

erw7 commented:
What is missing for this to be merged?

justinmk commented on test/functional/terminal/scrollback_spec.lua:
```diff
@@ -416,7 +440,12 @@ describe("'scrollback' option", function()
     command('sleep 100m')
-    feed_data('for i in $(seq 1 40); do echo "line$i"; done\n')
+    if iswin() then
+      nvim('command',
```
Can't feed_data be used ? It does jobsend on the child instance in the terminal.

erw7 committed:
Change to use feed_data

janlazo commented:
What's causing the tests to fail? There's no appveyor artifact to test.

erw7 committed:
Increase retry timeout

erw7 commented:
May be Screen:wait too slow in the CI environment. I tried retry timeout 10 times, passed the test.

janlazo commented:
`:terminal` works now and I can run fzf on it but on nvim-qt and the TUI, I have to press enter when fzf exits for the screen redraw on the terminal buffer. It doesn't return to the shell automatically.
Is this intended?

justinmk commented on test/functional/terminal/scrollback_spec.lua:
```diff
@@ -441,8 +439,7 @@ describe("'scrollback' option", function()
     if iswin() then
-      nvim('command',
-      'call jobsend(b:terminal_job_id, "for($i=1;$i -le 40;$i++){Write-Host \\"line$i\\"}\\<CR>")')
+      feed_data('for($i=1;$i -le 40;$i++){Write-Host \"line$i\"}\13') -- \13 = <CR>
```
Is there a reason `\n` can't be used instead of `\13`? It would be more consistent, since the non-windows case does that.

erw7 replied:
If I used ``` \n```, the command will not run. Therefore, I was using ```nvim('command', 'call jobsedn(...)")```.  If I used ```\13```, the command will run, from that I found out that I can use ```\r```. So I changed it to use ```\r```.

justinmk replied:
@erw7 Thanks, I was curious. I'm surprised that powershell doesn't accept `\n`.

erw7 committed:
Change from \13 to \r

erw7 commented:
I usually do not use fzf, but it doesn't matter where I tried a little. I tried with the following procedure in the following environment.

* Windows 7(64bit)
* cmd.exe
* fzf 0.16.7

```:term
> fzf
Press esc key
> fzf
Press return key
```

In both cases return to the shell without doing anything.

janlazo commented:
@erw7 Did you run fzf in a directory with some files and did you try running fzf in cmd.exe for comparison?
With no piped input, fzf in Windows runs [dir /s/b](https://github.com/junegunn/fzf/blob/d206949f6258db36060af3e84fdae15cb7b9a45c/src/constants_windows.go#L7) and outputs a selected path within the current directory when the user presses the return key. The escape key causes fzf to exit and output nothing.

One press of return/escape key should be enough for fzf to exit. `:term` requires 2 presses.

Environment
- Windows 8 (64 bit)
- nvim 0.2 (appveyor zip of this PR)
- fzf 0.16.x (tested `:term` with 0.16.0 and 0.16.7)
- cmd.exe or powershell.exe (tested both for comparison running native or `:term` and same result)
- TUI: `nvim.exe -u NONE`
- nvim-qt

erw7 commented:
Yes I run fzf in a directory with some files and also I tryed running fzf in cmd.exe for comparison. For the native cmd.exe and powershell.exe and for ```:set shell=cmd.exe``` ```: term``` and ```set shell=powershell.exe``` ```:term``` there is no difference. In all cases outputs the path selected with press the return key and fzf ends without outputting anything with press the escape key.

erw7 commented:
@janlazo Are you setting environment variables such as ```FZF_DEFAULT_COMMAND```, ```FZF_DEFAULT_OPTS``` etc..?

janlazo commented:
No.
I tried `FZF_DEFAULT_COMMAND` before in powershell to use ag, ack, etc. but it didn't work out because it left fzf hanging or fzf ran out of selections even if for simple queries in large directories.
I don't use `FZF_DEFAULT_OPTS` because I don't have a use case for it yet.
For `:term`, I don't export either environment variables and neither of them are set in cmd.exe and powershell.
`TERM` is unset for me because I'm not using sh/bash from cygwin.

erw7 commented:
Because I don't reproduce in my environment, I don't know the cause. But that is not intended.

janlazo commented on test/functional/terminal/scrollback_spec.lua:
```diff
@@ -416,7 +438,11 @@ describe("'scrollback' option", function()
     command('sleep 100m')
-    feed_data('for i in $(seq 1 40); do echo "line$i"; done\n')
+    if iswin() then
+      feed_data('for($i=1;$i -le 40;$i++){Write-Host \"line$i\"}\r')
```
Try `` `n `` for newline in powershell.
Use `` `r `` for carriage return and `` `r`n `` for carriage return + newline.

erw7 replied:
I wrote about this on the previous @justinmk point out. If I used \n, the command will not run. Is the command executed in the following procedure in your environment?

```
:call termopen(['powershell.exe'])
:call jobsend(b:terminal_job_id, "Get-ChildItem\n")
```

janlazo replied:
No, it's left hanging, waiting for more input (`>>` appears).

`\r` forces the command to run.
```viml
:call jobsend(b:terminal_job_id, "\r")
```

Seems `` `r `` and `` `n `` works only for multi-line output in `Write-Host`, `Write-Output`, etc.

erw7 replied:
@janlazo Thanks. Finally I understood the meaning you said. It was talking about whether the escape sequence of powershell is <code>&#x60;</code>, so <code>&#x60;n</code> can not be used. Since it is a special character of vim that has meaning here, it is no use to use the escape sequence of powershell. I am not confident as to whether I can explain it well because my poor English. However, I think that you can understand from the next execution result.

```
:echo map(split("\r`r", '\Z'), {idx, val -> char2nr(val)})
```

janlazo replied:
Got [13, 96, 114].
Get the char codes of carriage return, shell-specific escape character(s), and character(s) following them to resolve character combinations, such as `` `n ``, for terminal buffer.

Unrelated but, is something similar done to cmd.exe for newlines?

jalvesaq commented:
I tested the AppVeyor MINGW_64 binary on Windows, and Rterm.exe runs without issues. Even R's autocompletion works (R completes the names of objects from its workspace after the `TAB` key is pressed).

I might be confused, but as far as I remember, in the past, I had to call `jobsend(string)` on Windows finishing the string with `\r\n`. This is no longer necessary.

janlazo commented:
What happens if a program clears the screen on exit?

erw7 commented:
In my environment, ```vim``` and ```nvim``` have no particular difference compared to the native cmd.exe. When erasing the screen with the ```cls``` command, the buffer history is not deleted compared with the native cmd.exe.

janlazo commented:
`cls` in cmd.exe and `clear` in powershell behave as expected but they're commands.

[micro](https://github.com/zyedidia/micro/) requires 2 key presses of escape/enter to exit.
Both micro and fzf use [tcell](https://github.com/gdamore/tcell) and clear the screen on exit.

erw7 commented:
In my environment micro is exit with the ```Ctlr-q``` and no extra key input is required. I don't know golang well, but I watched little [console_win.go](https://github.com/gdamore/tcell/blob/master/console_win.go). I think tcell is calling the Windows API to clear the screen, and I think that it will not wait for key input by that process. 

I think that the possibility is low, but is something output to the file in the next procedure?

```
fzf 2> fzf.log
micro 2> micro.log
```

If application with full screen and prompting to stderr and waiting for input, it seems that sometimes the prompt is not displayed even with native cmd.exe.

janlazo commented:
I tested both with `nvim.exe -u NONE`. fzf.log and micro.log are empty files for me.
Exiting either with the escape key causes a delay but using enter key (or `\r` via `jobsend`), `Ctrl-q` (micro) is instant.

If this is an issue with the  OS (Win 7 vs Win 8.1) or golang,
I can live with it by using `:call jobsend(b:terminal_job_id, "\r")` when the program exits.

erw7 commented:
I installed Windows 10 on Virtualbox, and reproduced that issue. A similar issue also occurs with Cygwin's winpty.exe on Windows 10. Therefore, I think that is a issue of winpty. I have created an [issue for winpty](https://github.com/rprichard/winpty/issues/119).

justinmk committed:
Revert "Fix to fail build on *nix"

This reverts commit 73b24776a9995753ab6be0318a329ea6730437b6.

justinmk commented on src/nvim/CMakeLists.txt:
```diff
@@ -118,6 +118,10 @@ endforeach()
 
+if (NOT WIN32)
+  list(REMOVE_ITEM NVIM_HEADERS "${PROJECT_SOURCE_DIR}/src/nvim/os/pty_process_win.h")
+endif()
```
@erw7 Do you happen to remember why this was needed? I removed it in https://github.com/neovim/neovim/pull/7007/commits/b885b9358a4f1b0eb38a75b4a31839ba23346be3 and it seems to build OK on linux, bsd, and macOS.

justinmk replied:
I see now it fails the "single includes" test. Addressed that in 90de721d5ca1

erw7 replied:
Failed "single include test" as you say.
Will you work #7007 after this? Then I close this pull request.

justinmk replied:
@erw7 Yes, #7007 is almost ready, it's an effort to merge your work here.

justinmk committed:
single-includes: ignore os/pty_process_win.h

justinmk commented:
#7007 will merge this. Many thanks to @erw7 !
